src/Draco.Compiler/Internal/Symbols/TypeSymbol.cs

```diff
@@ -24,15 +25,91 @@ internal abstract partial class TypeSymbol : Symbol, IMemberSymbol
     /// </summary>
     public virtual bool IsValueType => false;
 
+    /// <summary>
+    /// True, if this type is interface.
+    /// </summary>
+    public virtual bool IsInterface => false;
+
     /// <summary>
     /// The substituted type of this one, in case this is a type variable.
     /// It's this instance itself, if not a type variable, or not substituted.
     /// </summary>
     public virtual TypeSymbol Substitution => this;
 
+    /// <summary>
+    /// The immediate base types of this type.
+    /// </summary>
+    public virtual ImmutableArray<TypeSymbol> BaseTypes => ImmutableArray<TypeSymbol>.Empty;
```

Comment:
If you rename this to `ImmediateBaseTypes` and add a `BaseTypes` that is all base types, you can get rid of a lot of recursion